Update EIP-2780: move state-dependent charges to runtime (cold surcharge, new-account state gas) (#11844)
ethereum/EIPs: EIPS/eip-2780.md

## Specification

### Parameters

| Name | Value | Definition |
| :----: | :----: | :---- |
| `TX_BASE_COST` | 12,000 | Sender cost: ECDSA recovery, the sender's account access, the sender's account write and inclusion in block |
| `TX_VALUE_COST` | 4,244 | Recipient balance write performed by a value transfer |
| `TRANSFER_LOG_COST` | 1,756 | [EIP-7708](./eip-7708.md) transfer log: `GAS_LOG + 3 × GAS_LOG_TOPIC + 32 × GAS_LOG_DATA_PER_BYTE` |

> The remaining values are defined in other EIPs and referenced here:

Reviewer: I am not sure about adding these numbers here. If we end up updating this values with other EIPs, then this EIP will become obsolete. I would propose keeping this as it was before

Author: Even previously, there were numbers here:
```
 `COLD_ACCOUNT_ACCESS = 3,000` and `CREATE_ACCESS = 11,000`, per [EIP-8038](./eip-8038.md).
- `STATE_BYTES_PER_NEW_ACCOUNT × CPSB = 183,600`, per [EIP-8037](./eip-8037.md). This value changes whenever a fork changes `CPSB`.
```
I just put it in a table and added the missing ones.

I dont think we can do a lot here, and it is easier to read this EIP with these values, and to have examples of consumption is necessary.

Reviewer: ok, good point. Let's leave the table then

| `ACCOUNT_WRITE` | 8,000 | [EIP-8038](./eip-8038.md) | First write to an account leaf |
| `CREATE_ACCESS` | 11,000 | [EIP-8038](./eip-8038.md) | Contract-deployment account access and write |
| `REGULAR_PER_AUTH_BASE_COST` | 7,816 | [EIP-8037](./eip-8037.md) | Per-authorization base: calldata, ECDSA recovery (per [EIP-7904](./eip-7904.md)), cold authority access, and warm writes |
| `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` | 183,600 | [EIP-8037](./eip-8037.md) | State gas for one new account leaf |
| `STATE_BYTES_PER_AUTH_BASE × CPSB` | 35,190 | [EIP-8037](./eip-8037.md) | State gas for one 23-byte delegation indicator |

The two state-gas products change whenever `CPSB` changes. Regular-gas and state-gas semantics are defined by [EIP-8037](./eip-8037.md).

### Intrinsic gas costs

The intrinsic gas is the state-independent cost of including the transaction in a block, computed without any state access, and is the sole input to the transaction-validity gas check. It is the sum of:

- `TX_BASE_COST` in regular gas — one ECDSA recovery, the sender's account access and write, and the inclusion of the transaction's bytes in the block — charged for every transaction;

> - if `tx.to` is not `None` and the transaction is not a self-transfer (`tx.sender != tx.to`), `COLD_ACCOUNT_ACCESS` in regular gas — the recipient touch, charged at the cold rate unconditionally;

Reviewer: I still prefer how the previous description was. I think it reads better and I don;t see why we need to change it.

Author: Yeah, the previous list is better, will revert.

- if the recipient is an [EIP-7702](./eip-7702.md) delegated account, additionally charge the delegation-target access in regular gas — `COLD_ACCOUNT_ACCESS` if cold, `WARM_ACCESS` if warm.

For a contract-creation transaction, the account at the deployment address is loaded instead; if it does not already exist, per the existence rule of [EIP-8037](./eip-8037.md), charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas. Whether the target exists can only be determined by reading state, so the check is part of the runtime phase.

The first frame is then entered with the gas remaining after these runtime charges.

Running out of runtime gas does **not** invalidate the transaction. Validity is decided solely by the intrinsic gas check, which needs no state access; the runtime charges, which do, can never retroactively invalidate an already-validated transaction. On out-of-gas the transaction is included, halts, the sender pays for all gas consumed, and all state changes — including any [EIP-7702](./eip-7702.md) delegations applied by its authorizations — are reverted, exactly as for an out-of-gas inside any call frame.

> If a contract-creation transaction's initcode reverts, the transaction remains valid and any `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` charged at runtime is refilled in last-in, first-out (LIFO) order, per [EIP-8037](./eip-8037.md): the portion the charge drew from `gas_left` is credited back to `gas_left`, and only the remainder returns to the `state_gas_reservoir`. On an exceptional halt the same refill applies, but `gas_left` is consumed, so only the reservoir portion survives. A plain value transfer to a new account executes no code and so cannot revert; its new-account state charge is therefore never rolled back this way.

Reviewer 2: > On an exceptional halt the same refill applies, but `gas_left` is consumed, so only the reservoir portion survives.

It confuses me.

For example, if we have a contract creation transaction, and the top-level create frame halts (e.g. due to an address collision). Should we also burn the transaction's remaining regular gas after applying the LIFO refunds?

For comparison, for `CREATE/CREATE2`, if the child create frame halts, we:

- revert its state changes and refund the child frame's state cost via LIFO (resetting the child's reservoir back to its original value);
- refund the parent frame's state cost via LIFO.

The parent frame then continues execution normally; it is not itself considered halted.

My question is whether a top-level contract creation transaction should behave differently. Should the top-most frame also preserve its remaining regular gas after the LIFO refunds, or should the halt be treated as propagating to the top level, causing the remaining regular gas to be burned as a special case?

cc @misilva73

Reviewer 2: > A plain value transfer to a new account executes no code and so cannot revert; its new-account state charge is therefore never rolled back this way.

It's not true. In the spec test, there is a test case that:
- transaction carries value
- to was not existent
- to is a precompile (0x07)
- precompile execution is failed due to invalid parameter
- the call frame is halted

In this special case, the account-creation will be reverted

The same question, in this special case, after applying the state refund with LIFO, should the halt be treated as propagating to the top level, causing the remaining regular gas to be burned as a special case?

@misilva73

Reviewer: On the first point, we should burn the `gas_left` as we do today in both CREATE and create transactions. Or am I missing something?

Also, good point with the precompile edge-case. We should clarify this.

Author: Yeah, no special case, it should behave the same as in frame,

Author: There was a bug with state clear and empty precompile in the past, so all precompiles now have at least some balance. Dont mind adding this to EIP but just saying

This EIP decomposes the flat `21,000` intrinsic transaction cost into explicit primitives, each priced to match the resources a transaction actually consumes:

> - `TX_BASE_COST = 12,000` — ECDSA recovery, the sender account's access and write, and the inclusion of the transactio in the block (which is transient and expires with history).

Reviewer 2: transactio typo

> - `TX_BASE_COST = 12,000` — ECDSA recovery, the sender account's access and write, and the inclusion of the transactio in the block (which is transient and expires with history).

Author: ```suggestion
- `TX_BASE_COST = 12,000` — ECDSA recovery, the sender account's access and write, and the inclusion of the transaction in the block (which is transient and expires with history).
```

> - `TX_BASE_COST = 12,000` — ECDSA recovery, the sender account's access and write, and the inclusion of the transaction in the block (which is transient and expires with history).

Reviewer 3: I think this should note that ECDSA recovery is 3000 gas, the sender account's access and write (what is this? `COLD_ACCOUNT_ACCESS` and `ACCOUNT_WRITE`, taken from EIP-8038?)

> and the inclusion of the transaction in the block (which is transient and expires with history).

How much is this cost?

Reviewer: Very early on, we decided to set the `TX_BASE_COST` based on benchmarks because otherwise there is not way for us to be below 21k for simple ETH transfers. ECDSA recovery at 3000 is overpriced, but changing the costs of this opcode is out of scope, so we decided this was the best way to approach this EIP.

Author: I guess this is compensated as can't be the same as in frame. For starters, ecdsa is done in parallel for all tx, so it can't be 3k gas. Caller and target are known and can be preloaded etc. I think this value is observed from benchmark, so it is fine to not specify EIP-8038 as they don't apply.

Reviewer 4: @misilva73 In that case, I don't think we should settle on a calldata floor too early. From what I can see, the proposed EIP-8311 is deriving its calldata pricing based on this 21k baseline.

- `COLD_ACCOUNT_ACCESS = 3,000` — the recipient account touch, per [EIP-8038](./eip-8038.md).
- `TX_VALUE_COST = 4,244` — the recipient balance write performed by a value transfer.
- `TRANSFER_LOG_COST = 1,756` — the [EIP-7708](./eip-7708.md) transfer log.

A plain ETH transfer to an existing account still costs `12,000 + 3,000 + 4,244 + 1,756 = 21,000` gas. The headline number is unchanged, but it is now built from measured components rather than a single legacy constant, so each transaction pays for the work it does and nothing it does not.

> The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.

Reviewer 3: If I understand it correctly I think we can simplify this change.

```suggestion
The gas charging is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered from their respective gas dimension. In case that the transaction cannot pay for these costs, all the remaining gas is consumed and all state changes are reverted.
```

> all state changes are reverted

is somewhat confusing, because the account nonce and the account balance (+coinbase balance if tx has priority fee) always changes, so that obviously should not be reverted.

The current text says: 

> as for any out-of-gas halt inside a frame.

This is not true, if the execution part of the tx OOGs currently, then all state changes in execution are rolled back, but the EIP-7702 delegations still persist. In this situation (insufficient gas to pay for all the state changes) it is different, as prior to entering the first frame we already OOG. 

Currently (Osaka) the state creation gas for EIP-7702 authorizations is guaranteed to exist because it comes from the 25k intrinsic cost of a EIP-7702 transaction (get a refund if the account exists). However, now once we enter the frame we start charging state gas costs and also regular gas costs. The question is: in which order is this charged? What if we can pay state gas for a few EIP-7702 authorizations, but for others not? Do we rollback all authorizations we just made? Or do we keep those we can pay for and then OOG? This should be addressed. In case that we keep the ones we can pay for then it must be made clear in what specific order these gas costs are deducted, as this in some situations influences if an authorization could be paid for or not.

Reviewer 3: This abstract is too large and already specifies some constants and describes behavior, the abstract should summarize what this EIP does and the respective details go in the relevant sections.

> The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.

Reviewer 5: ```suggestion
The gas charging is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.
```

> The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.

Reviewer 5: ```suggestion
The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in the same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.
```

> The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction it is considered as halt. Transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.

Reviewer: ```suggestion
The gas charing is split across two phases. **Intrinsic gas** is the state-independent cost of including the transaction in a block and is the sole input to the transaction-validity check. **Runtime gas** covers the state-dependent charges — the recipient cold surcharge, new-account creation, and delegation resolution - charged as the first frame is entered, metered in same way as inside the frame. Running out of gas in runtime does not invalidate the transaction. Instead, the transaction is included, execution is skipped, and all state changes are reverted, as for any out-of-gas halt inside a frame.
```

This EIP does not change calldata or access-list metering.

## Motivation

The flat `21,000` intrinsic cost charges every transaction the same amount regardless of what it does. A zero-value transaction to an existing account and a value transfer that creates a new account both pay `21,000`, even though the latter touches an extra account, writes an extra balance, emits a log, and permanently grows the state. The single constant both over-prices the simplest paths and under-prices the ones that consume real, lasting resources.

This EIP replaces the constant with a decomposition into named primitives, each priced to the resource it represents. The goal is to make every transaction's cost reflect its actual work.

> Transaction costs fall into two groups. Some can be computed from the transaction's fields alone, such as signature recovery, the sender's account access and write, and the transaction's bytes in the block. Others require reading state, such as touching the recipient, creating a new account, or resolving a delegation. This EIP splits the two: intrinsic gas is reduced to the minimum needed to include the transaction in a block and remains the sole input to the validity check, while state-dependent costs are charged at runtime, using the same metering as inside a frame. This removes the overcharging of the flat worst-case reservation, since a transaction pays a state-dependent cost only when it actually incurs it. It also allows halting in the pre-execution phase: a transaction whose gas covers inclusion but not a state-dependent charge is included, charged for the gas consumed, and reverted, exactly like an out-of-gas inside a call frame.

Reviewer 3: > the transaction's bytes in the block

How does this compute work? Is this the payment for the tx-bytes which are not the data field (as that is paid for as calldata cost)?

Author: Yes, it is data fields and access list and auth list.

This EIP replaces the constant with a decomposition into named primitives, each priced to the resource it represents. The goal is to make every transaction's cost reflect its actual work.

Transaction costs fall into two groups. Some can be computed from the transaction's fields alone, such as signature recovery, the sender's account access and write, and the transaction's bytes in the block. Others require reading state, such as touching the recipient, creating a new account, or resolving a delegation. This EIP splits the two: intrinsic gas is reduced to the minimum needed to include the transaction in a block and remains the sole input to the validity check, while state-dependent costs are charged at runtime, using the same metering as inside a frame. This removes the overcharging of the flat worst-case reservation, since a transaction pays a state-dependent cost only when it actually incurs it. It also allows halting in the pre-execution phase: a transaction whose gas covers inclusion but not a state-dependent charge is included, charged for the gas consumed, and reverted, exactly like an out-of-gas inside a call frame.

> - **Accuracy.** Signature recovery, account accesses, account writes, the transfer log, and state creation are billed separately, so cost tracks usage.

Reviewer: Why are we removing this part?

Author: Can revert

Reviewer: Please do

Reviewer: And, I would actually move the new paragraph above as a new bullet point in this list.

Author: Reverted

| Name | Value | Definition |
| :----: | :----: | :---- |
| `TX_BASE_COST` | 12,000 | Sender cost: ECDSA recovery, the sender's account access, the sender's account write and inclusion in block |

> | `TX_VALUE_COST` | 4,244 | Recipient balance write performed by a value transfer |

Reviewer 3: This constant should be expressed in EIP-8038 terms, right? In case of value sent, charge extra `ACCOUNT_WRITE`?

Author: They are not the same constants as some are compensated so we can't use `ACCOUNT_WRITE`

Reviewer: Yes, not the same constants and with different benchmarked values

> | `TRANSFER_LOG_COST` | 1,756 | [EIP-7708](./eip-7708.md) transfer log: `GAS_LOG + 3 × GAS_LOG_TOPIC + 32 × GAS_LOG_DATA_PER_BYTE` |

Reviewer 3: Not relevant for this PR but EIP-7708 does not seem to charge for the introduces logs, so I dont see why we should charge it here.

Author: cc @misilva73, maybe worth clarifying in EIP-8038 or EIP-7708

Reviewer: This was a proposal from @benaadams . I think we should maybe add this to 7708 to make it consistent

The remaining values are defined in other EIPs and referenced here:

| Name | Value | Source | Definition |
| :---- | :----: | :----: | :---- |
| `COLD_ACCOUNT_ACCESS` | 3,000 | [EIP-8038](./eip-8038.md) | Cold account touch |
| `WARM_ACCESS` | 100 | [EIP-8038](./eip-8038.md) | Warm account touch |
| `ACCOUNT_WRITE` | 8,000 | [EIP-8038](./eip-8038.md) | First write to an account leaf |
| `CREATE_ACCESS` | 11,000 | [EIP-8038](./eip-8038.md) | Contract-deployment account access and write |

> | `REGULAR_PER_AUTH_BASE_COST` | 5,657 | [EIP-8037](./eip-8037.md) | Per-authorization base: calldata, ECDSA recovery (per [EIP-7904](./eip-7904.md)), warm authority access, and warm writes |

Reviewer 5: The per-authorization base fee of 5,657 already contains a full cold account access (per EIP-8037), so adding another cold cost at runtime makes cold authorities pay for the same access twice.

Author: After discord discussion, we will always charge cold gas.

> | `REGULAR_PER_AUTH_BASE_COST` | 5,657 | [EIP-8037](./eip-8037.md) | Per-authorization base: calldata, ECDSA recovery (per [EIP-7904](./eip-7904.md)), warm authority access, and warm writes |

Reviewer 3: How is this 5657 number calculated? Seems to be the sum of the component costs?

Author: I think it is benchmarking

| `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` | 183,600 | [EIP-8037](./eip-8037.md) | State gas for one new account leaf |
| `STATE_BYTES_PER_AUTH_BASE × CPSB` | 35,190 | [EIP-8037](./eip-8037.md) | State gas for one 23-byte delegation indicator |

> The two state-gas products change whenever a fork changes `CPSB`. Regular-gas and state-gas semantics are defined by [EIP-8037](./eip-8037.md).

Reviewer 3: ```suggestion
The two state-gas products change whenever `CPSB` changes. Regular-gas and state-gas semantics are defined by [EIP-8037](./eip-8037.md).
```

This is to make this EIP forward-compatible in case we introduce something like dynamic CPSB.

### Intrinsic gas costs

The intrinsic gas is the state-independent cost of including the transaction in a block, computed without any state access, and is the sole input to the transaction-validity gas check. It is the sum of:

> - `TX_BASE_COST` in regular gas — one ECDSA recovery, the sender's account access and write, and the inclusion of the transaction's bytes in the block — charged for every transaction;

Reviewer 3: I would like to see a full breakdown on how this evaluates to 12k, where it is possible using EIP-8038 constants.

Reviewer: There is no breakdown. This comes directly from benchmarks. Check the rationale section for more details

Author: You can't see that, and EIP-8038 does not translate here. Doing math on eip8038 does not work here as those values are for inside frame, for transaction to/target a lot is done before block even starts executing (for example parallel recovery) so it is compensated.

The value of 12k is obtained from benchmarking, iirc.

> - if `tx.to` is not `None` and the transaction is not a self-transfer (`tx.sender != tx.to`), `WARM_ACCESS` in regular gas — the state-independent floor of the recipient touch; the cold surcharge, if any, is charged at runtime;

Reviewer: I don't think we should charge `WARM_ACCESS` here and then adjust to `COLD_ACCESS_COST`. We should always charge `COLD_ACCESS_COST`. In fact, I don't see why we are changing so much this part of the spec.

From my understanding, the only change should be delaying the `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` charge for contract deployments.

Reviewer 4: > I don't think we should charge `WARM_ACCESS` here and then adjust to `COLD_ACCESS_COST`. We should always charge `COLD_ACCESS_COST`. In fact, I don't see why we are changing so much this part of the spec.

So we should overcharge and then refund if account is warm

Author: We will always charge, and skip the refund

- if `tx.to` is `None` (contract-creation transaction), `CREATE_ACCESS` in regular gas; the new-account state charge depends on whether the deployment target exists and is charged at runtime;
- value charges, from `tx.value` and whether the transaction is a self-transfer (`tx.sender == tx.to`):
  - if `tx.value == 0`, or the transaction is a self-transfer, no charge;
  - if `tx.value > 0` and contract creation, `TRANSFER_LOG_COST` in regular gas;

>   - otherwise, `TRANSFER_LOG_COST + TX_VALUE_COST` in regular gas;

Reviewer 3: I dont think we should charge the logs. I think we can summarize that `TX_VALUE_COST` is paid if value > 0 and the target is not self. 

In case that we create a contract, target is per definition not self so we pay for it.

We should not pay for these logs because EIP-7708 also does not incur costs for them.

Reviewer: Again, this was proposed by @benaadams. And we should update 7708 accordingly. Happy to discuss this more, if you feel strongly about not including it.

Author: iirc this value is obtained by benchamrking

- `REGULAR_PER_AUTH_BASE_COST` in regular gas per [EIP-7702](./eip-7702.md) authorization;
- calldata and access-list costs, unchanged.

> Only the warm floor of the recipient touch is intrinsic; the cold surcharge (`COLD_ACCOUNT_ACCESS − WARM_ACCESS`) is state-dependent and is charged at the runtime phase below. These charges replace the legacy contract-creation transaction charges and EIP-7702's flat `PER_EMPTY_ACCOUNT_COST × authorization_list_length` intrinsic charge. Calldata and access-list metering remain unchanged.

Reviewer 3: Why is the cold surcharge state-dependent? I have to cold charge if recipient is not self - right? This is not state dependent. If we charge an extra `COLD_ACCOUNT_ACCESS` if we hit an EIP-7702 delegated account, then that would be state dependent.

> These charges replace the legacy contract-creation transaction charges and EIP-7702's flat `PER_EMPTY_ACCOUNT_COST × authorization_list_length` intrinsic charge. 

So how does EIP-7702 intrinsic gas change? Does the intrinsic gas now change to `COLD_ACCOUNT_ACCESS x authorization_list_length`?

Author: I'll modify this, from discord discussion it now should charge COLD+recovery in intrinsic

Author: Ok, already modified

### Runtime gas costs

Once the transaction passes the intrinsic gas check, the state-dependent charges are applied at runtime — as the first frame is entered — drawing from the same gas the transaction supplies. Like every charge in this EIP, each runtime charge is split into regular gas and state gas per [EIP-8037](./eip-8037.md): account accesses and writes are regular gas, while durable state growth (new account leaves and net-new delegation bytes) is state gas. This mirrors how a `CALL` frame is metered: accesses, account creation, and writes are charged as state is touched, not up front. Because these charges require state access, they are evaluated only after [EIP-7702](./eip-7702.md) authorizations are applied.

While processing each valid authorization — the per-authority rules of [EIP-8037](./eip-8037.md) ensure each charge below is applied at most once per authority:

- if the authority address is cold — not already in `accessed_addresses`, per [EIP-2929](./eip-2929.md) — charge the cold surcharge `COLD_ACCOUNT_ACCESS − WARM_ACCESS` in regular gas;

> - if the authority did not exist, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas and `ACCOUNT_WRITE` in regular gas for the new account leaf;

Reviewer 2: It's not true for AccountWrite.

The AccountWrite should be charged for any valid authorization, regardless of the authority was existent or not, or the delegation net change is empty or not.

For any valid authorization, the nonce of authority will be bumped and database write will happen anyway.

Reviewer 2: Let's extend it with one more rule:

```
- if the authority account does not already exist, charge STATE_BYTES_PER_NEW_ACCOUNT × CPSB in state gas for creating the new account leaf;
- if this is the first update to the authority within the transaction(    differs from tx.to), charge ACCOUNT_WRITE in regular gas for updating the account.
```

Author: Thanks, have used those rules

Reviewer 2: Just realize the authority can also be tx.from.

We should improve it with

`- if this is the first update to the authority within the transaction(e.g. differ from tx.from and tx.to), charge ACCOUNT_WRITE in regular gas for updating the account.`

> - if the authorization writes the 23-byte delegation indicator into a previously empty slot (net-new delegation bytes), charge `STATE_BYTES_PER_AUTH_BASE × CPSB` in state gas.

Reviewer 3: What is meant by previously empty slot? Do you mean it has no code? What if I set a delegation, then set it to zero address (remove the delegation) and then set it again, do we charge this gas or not? It is not clear to me what is meant with the `net-new`, it seems like we should charge this twice.

`REGULAR_PER_AUTH_BASE_COST` prices the authority access at the warm rate; the cold surcharge above is the only account-access charge added at runtime for the authority.

> Then the recipient account is loaded, and unless the transaction is a self-transfer or a contract creation:

Reviewer 5: What if the self-transfer is to myself but I have a delegation set, this cannot just cost 12k

Author: It will cost an additional cold load for delegates.

- if the recipient is cold — not warmed by an applied authorization or an access list, per [EIP-2929](./eip-2929.md) — charge the cold surcharge `COLD_ACCOUNT_ACCESS − WARM_ACCESS` in regular gas; the warm floor is already covered by the intrinsic `WARM_ACCESS`;
- if the recipient is non-existent and `tx.value > 0`, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas;
- if the recipient is an [EIP-7702](./eip-7702.md) delegated account, additionally charge the delegation-target access in regular gas — `COLD_ACCOUNT_ACCESS` if cold, `WARM_ACCESS` if warm.

> For a contract-creation transaction, the account at the deployment address is loaded instead; if it does not already exist, per the existence rule of [EIP-8037](./eip-8037.md), charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas. Whether the target exists can only be determined by reading state, so the check is part of the runtime phase.

Reviewer 5: This conflicts with 8037:
8037 says to chargs upfront and give it back afterwards if target existed, and this pr would first look at the target and only charge if it doesn't exist.

Author: True, i maybe rushed it. @chfast did propose to change this behaviour [here](https://github.com/ethereum/EIPs/pull/11858)

Will revert back to how it is currently specified in EIP-8037

Reviewer 3: Flagging that this conflict still exists. I think part of confusion and contradictions come from the fact that both EIPs depend on each other (EIP-8037 depends on EIP-2780, and vice-versa), which makes it unclear what EIP should specify *what* exactly.

> For a contract-creation transaction, the account at the deployment address is loaded instead; if it does not already exist, per the existence rule of [EIP-8037](./eip-8037.md), charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas. Whether the target exists can only be determined by reading state, so the check is part of the runtime phase.

Reviewer 3: Also pay `CREATE_ACCESS` regular gas?

Author: This is part of the runtime cost, while `CREATE_ACCESS` is an intrinsic cost, so it makes sense it is not here.

The first frame is then entered with the gas remaining after these runtime charges.

> Running out of runtime gas does **not** invalidate the transaction. Validity is decided solely by the intrinsic gas check, which needs no state access; the runtime charges, which do, can never retroactively invalidate an already-validated transaction. On out-of-gas the transaction is included, halts, the sender pays for all gas consumed, and all state changes — including any [EIP-7702](./eip-7702.md) delegations applied by its authorizations — are reverted, exactly as for an out-of-gas inside any call frame.

Reviewer 3: This changes behavior if we compare it to txs which currently OOG once it executes a contract. In that case, all state changes made during the contract execution are reverted, but all EIP-7702 authorizations are kept. If this behavior change is intended then this should be noted.

Author: Should clarify, only if pre execution halts, auth list is reverted. If call frame reverts auth list stays.

> If a contract-creation transaction's initcode reverts, the transaction remains valid and any `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` charged at runtime is refilled to the `state_gas_reservoir`, per [EIP-8037](./eip-8037.md). A plain value transfer to a new account executes no code and so cannot revert; its new-account state charge is therefore never rolled back this way.

Reviewer 3: So we go OOG, but then get this `STATE_BYTES_PER_NEW_ACCOUNT x CPSB` gas back? So lets say I send Y gas limit which OOGs pre-execution, since I could not pay for all of the costs combined. But my final gas usage will be `Y - STATE_BYTES_PER_NEW_ACCOUNT * CPSB` since I get this "refunded" to the state reservoir and which I thus did not spend / consume? 🤔 

Also want to note: this new pre-execution OOG situation is rather nasty, because it is not directly visible if the OOG happened pre-execution or during the execution 🤔

Author: OOG consumes all the gas, there is no gas left (This is how the latest EIP-8037 behaves).

### Transaction reference cases

Costs are split between the intrinsic phase and the runtime phase, each in regular gas and state gas, per [EIP-8037](./eip-8037.md). "execution" denotes contract execution charged at the standard schedule; it, and every runtime charge, is drawn from the gas passed into the first frame, so that frame is entered with the gas remaining after the runtime regular and state charges. The recipient's warm floor `WARM_ACCESS` is intrinsic; its cold surcharge `COLD_ACCOUNT_ACCESS − WARM_ACCESS` and any delegation-target access are runtime. Cases show the cold rate; a recipient warmed beforehand pays no runtime surcharge, and a delegation target warmed beforehand pays `WARM_ACCESS` rather than `COLD_ACCOUNT_ACCESS`. Create rows assume the deployment target does not already exist; a pre-existing target pays no new-account state charge. The intrinsic + runtime totals match the legacy headline numbers.

| Case | Intrinsic (regular / state) | Runtime (regular / state) | Total (regular / state) |
| :---- | :---- | :---- | :----: |

> | ETH transfer to self | `TX_BASE_COST` / 0 | 0 / 0 | 12,000 / 0 |

Reviewer 3: This table is super handy!

| Create transaction, `tx.value` > 0, `created_address.balance` > 0 | `TX_BASE_COST + CREATE_ACCESS + TRANSFER_LOG_COST` / 0 | 0 / 0 | 24,756 / 0 |

### Interactions with other EIPs

- **[EIP-2929](./eip-2929.md) (warm/cold accounting).** The sender access folded into `TX_BASE_COST` is always charged at the intrinsic phase, as is the recipient's `WARM_ACCESS` floor. The recipient cold surcharge (`COLD_ACCOUNT_ACCESS − WARM_ACCESS`), the authority cold surcharge, and the delegation-target access are charged at the runtime phase and follow the standard EIP-2929 warm/cold model: a recipient or authority already warmed — by an access list or by an [EIP-7702](./eip-7702.md) authorization applied earlier in the same transaction — incurs no cold surcharge, and a warm delegation target pays `WARM_ACCESS` rather than `COLD_ACCOUNT_ACCESS`. The same model governs all execution-level account touches, including internal `CALL`s.
- **[EIP-2930](./eip-2930.md) (access lists).** Access lists keep their existing per-entry charges and warming semantics for execution-level touches.

> - **[EIP-7623](./eip-7623.md) (calldata floor).** EIP-7623 floors a transaction's cost at `21,000 + TOTAL_COST_FLOOR_PER_TOKEN × tokens_in_calldata`, where the flat `21,000` is the intrinsic base this EIP decomposes. With this EIP active, that base term is replaced by the transaction's decomposed regular-gas intrinsic — the sum of the regular-gas intrinsic primitives (`TX_BASE_COST + WARM_ACCESS + TX_VALUE_COST + TRANSFER_LOG_COST` for value transfers, `TX_BASE_COST + CREATE_ACCESS` for contract creation) — so the floor and its validity check rest on the same per-transaction base as the rest of intrinsic gas rather than a stale constant. Like the floor, this base is state-independent, so the state-dependent runtime charges (the recipient cold surcharge and the new-account state-gas charge) do not enter it. The calldata schedule itself is unchanged; only the base the floor sits on moves.

Reviewer 3: I think https://eips.ethereum.org/EIPS/eip-3651 (warm coinbase) should be included here as well to note that coinbase is thus also warm (for completeness) and should also be added as required EIP if it is not added.

Author: This got changes after discord discussion, and cold cost is part of TX_BASE

- **[EIP-7708](./eip-7708.md) (ETH transfers emit a log).** Every non-zero-value transfer to a different account emits a transfer log, priced explicitly as `TRANSFER_LOG_COST` rather than absorbed into the base.

> - **[EIP-7928](./eip-7928.md) (block-level access lists).** `tx.sender` is accessed at the intrinsic phase and `tx.to` at the runtime phase, so both are included in the BAL even when the transaction reverts.

Reviewer 3: This yields a lot of edge cases, because we now require the ordering of all gas operations and a very clear definition when state is accessed or not. For contract create transactions what do we charge first, state or regular gas? (when do we perform the check if the account exists or not? This is the moment to include it as "read" in the BAL)

Author: This is not a lot different from things that we do inside frame. We should follow same rules that we have there.

I should clarify it in text more. Order of operations is already set, auth list first than `tx.to` so if we halt in auth list, `tx.to` should not be part of BAL.

If a contract-creation transaction's initcode reverts, the transaction remains valid and any `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` charged at runtime is refilled in last-in, first-out (LIFO) order, per [EIP-8037](./eip-8037.md): the portion the charge drew from `gas_left` is credited back to `gas_left`, and only the remainder returns to the `state_gas_reservoir`. On an exceptional halt the same refill applies, but `gas_left` is consumed, so only the reservoir portion survives. A plain value transfer to a new account executes no code and so cannot revert; its new-account state charge is therefore never rolled back this way.

### Transaction reference cases

Costs are split into regular gas and state gas, per [EIP-8037](./eip-8037.md). "execution" denotes contract execution charged at the standard schedule.

| Case | Regular gas | State gas | Regular gas value | State gas value |
| :---- | :---- | :---- | :----: | :----: |

> | ETH transfer to self | `TX_BASE_COST` | 0 | 12,000 | 0 |

Reviewer 4: Why isn't there a case covering a self-transfer to an EIP-7702 delegated account (`tx.sender == tx.to` and `tx.to` contains EIP-7702 delegation code)?

I would expect the cost to be the same as in the first case, except for the additional cost of accessing the delegated address.

### Runtime gas costs

Once the transaction passes the intrinsic gas check, the state-dependent charges are applied at runtime — as the first frame is entered — drawing from the same gas the transaction supplies. Like every charge in this EIP, each runtime charge is split into regular gas and state gas per [EIP-8037](./eip-8037.md): account accesses and writes are regular gas, while durable state growth (new account leaves and net-new delegation bytes) is state gas. This mirrors how a `CALL` frame is metered: accesses, account creation, and writes are charged as state is touched, not up front. Because these charges require state access, they are evaluated only after [EIP-7702](./eip-7702.md) authorizations are applied.

While processing each valid authorization — the per-authority rules of [EIP-8037](./eip-8037.md) ensure each charge below is applied at most once per authority:

- if the authority account does not already exist, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas for the new account leaf;

> - if this is the first write to the authority within the transaction (i.e. the authority differs from `tx.to`), charge `ACCOUNT_WRITE` in regular gas;

Reviewer 6: Is this supposed to be `tx.sender` instead of `tx.to`?

Reviewer 2: See https://github.com/ethereum/EIPs/pull/11891

Then the recipient account is loaded and added to `accessed_addresses` — its access was already charged at the cold rate at the intrinsic phase — and unless the transaction is a self-transfer or a contract creation:

- if the recipient is non-existent and `tx.value > 0`, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas;
- if the recipient is an [EIP-7702](./eip-7702.md) delegated account, additionally charge the delegation-target access in regular gas — `COLD_ACCOUNT_ACCESS` if cold, `WARM_ACCESS` if warm.

For a contract-creation transaction, the account at the deployment address is loaded instead; if it does not already exist, per the existence rule of [EIP-8037](./eip-8037.md), charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas. Whether the target exists can only be determined by reading state, so the check is part of the runtime phase.

The first frame is then entered with the gas remaining after these runtime charges.

> Running out of runtime gas does **not** invalidate the transaction. Validity is decided solely by the intrinsic gas check, which needs no state access; the runtime charges, which do, can never retroactively invalidate an already-validated transaction. On out-of-gas in this pre-execution phase the transaction is included, halts before the first frame is entered, the sender pays for all gas consumed, and all state changes are reverted — including any [EIP-7702](./eip-7702.md) delegations, which an out-of-gas *inside* the frame would leave in place, per [EIP-7702](./eip-7702.md).

Reviewer 6: Think the following could be a bit confusing and might benefit from further clarification. From what I can tell now, there are 3 possible exceptional halts.

1. While setting the delegation for a type 4 transaction at the run-time phase
2. While charging the additional state related costs at the run-time phase (value transfer to a non-existent account/7702 delegation access)
3. Exceptional halt while executing the opcodes in the top frame

From what I understand, only the run time exceptional halts (1 and 2) should revert the state changes. 3 should not. Might be worth making that explicit.

Reviewer 6: The reason why this sticks out a bit is because elsewhere there is a line that says "Once the transaction passes the intrinsic gas check, the state-dependent charges are applied at runtime — as the first frame is entered — drawing from the same gas the transaction supplies." Reading these two side-by-side, feels like they contradict on where the "first frame" begins

Reviewer: I agree that it makes more sense to only revert in cases 1 and 2. We can however merge this PR and I can add this clarification in the clean-up

### Decomposing authorization cost

[EIP-7702](./eip-7702.md) charges `PER_EMPTY_ACCOUNT_COST` per authorization at the intrinsic phase and refunds it when the authority already exists, over-reserving gas the transaction may never use. Splitting the cost removes the over-charge: only `REGULAR_PER_AUTH_BASE_COST` — the calldata, signature recovery, warm authority access, and warm writes every authorization performs — is knowable without state access and stays intrinsic. The cold surcharge for a cold authority, the new-account leaf and its `ACCOUNT_WRITE`, and the delegation-indicator bytes depend on state and are charged at runtime during authorization processing, billed only for the authorities that incur them.

## Backwards Compatibility

This EIP is **not** backward compatible. It is a consensus gas repricing that must be activated at a network upgrade. Wallets, RPCs, gas estimators, and any logic that assumes a flat `21,000` intrinsic base must update, particularly to account for the state-gas charge on transactions that create new accounts.

## Test Cases

Costs are given as intrinsic + runtime, by case:

1. Self-transfer (`from == to`): `12,000` intrinsic, `0` runtime.

> 2. Zero-value transaction to any address (existing, empty, or contract): `12,100` intrinsic (`TX_BASE_COST + WARM_ACCESS`) + `COLD_ACCOUNT_ACCESS − WARM_ACCESS` runtime = `15,000` regular, `0` state. A self-target is `12,000` with no runtime charge.

Reviewer 5: ```suggestion
2. Zero-value transaction to a cold recipient (existing, empty, or contract): 12,100 intrinsic (TX_BASE_COST + WARM_ACCESS) + COLD_ACCOUNT_ACCESS − WARM_ACCESS runtime = 15,000 regular, 0 state. If the recipient is already warm, a precompile, the COINBASE (per EIP-3651), or an EIP-2930 access-list entry, the runtime cold surcharge does not apply: 12,100 regular, 0 state. A self-target is 12,000 with no runtime charge.
```

9. Contract-creation transaction whose initcode reverts (value-bearing or not): the transaction remains valid; the `183,600` new-account state-gas charge is refilled to the reservoir.
10. Transaction whose gas covers intrinsic cost but not a runtime charge — e.g. a zero-value call funded for its intrinsic gas (`TX_BASE_COST + WARM_ACCESS`) but not the recipient cold surcharge (`COLD_ACCOUNT_ACCESS − WARM_ACCESS`), a value transfer to a non-existent recipient or a contract creation to a non-existent target lacking gas for the new-account state charge, or a transfer to a delegated account lacking gas for the delegation-target access: the transaction is valid and included, halts out-of-gas, and all runtime state changes — including any applied EIP-7702 delegations — are reverted; it is **not** invalidated.

Blocks of pure ETH transfers should be tested across all EL clients (e.g. on Perfnet) to confirm the pricing matches measured runtimes at the chosen throughput anchor.

## Security Considerations

> The decomposed charges, intrinsic plus runtime, sum to at least the legacy `21,000` for any transaction that moves value to a distinct account, and add state gas for transactions that grow the state. No path is cheaper than under the flat base in a way that increases per-block resource consumption; the change reallocates cost to match resources and prices state growth that was previously unpriced.

Reviewer 5: ```suggestion
The decomposed charges, intrinsic plus runtime, sum to at least the legacy `21,000` for any transaction that moves value to a distinct, cold account, and add state gas for transactions that grow the state. No path is cheaper than under the flat base in a way that increases per-block resource consumption; the change reallocates cost to match resources and prices state growth that was previously unpriced.
```

Reviewer 5: with 7702 the account could be warmed already then the "at least" doesn't apply anymore
